js/src/dom/manipulator.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * --------------------------------------------------------------------------
+ * Bootstrap (v4.3.0): dom/manipulator.js
+ * Licensed under MIT (https://github.com/twbs/bootstrap/blob/master/LICENSE)
+ * --------------------------------------------------------------------------
+ */
+
+const regexDataKey = /[A-Z]/g
+
+function normalizeData(val) {
+  if (val === 'true') {
+    return true
+  } else if (val === 'false') {
+    return false
+  } else if (val === 'null') {
+    return null
+  } else if (val === Number(val).toString()) {
+    return Number(val)
+  } else if (val === '') {
+    return null
+  }
+
+  return val
+}
+
+function normalizeDataKey(key) {
+  return key.replace(regexDataKey, (chr) => chr.toLowerCase())
+}
+
+const Manipulator = {
+  setDataAttribute(element, key, value) {
+    element.setAttribute(`data-${normalizeDataKey(key)}`, value)
```

Comment:
Can't `element.dataset` be used here? It's supported in all browsers: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset

Reply:
Seems not, for Safari it's unknown 😟

Reply:
Hmm that's odd, I just tested with Browserstack all the way to Mac Safari 5.1.9 (6534.59.8) (used on OSX Snow Leopard, released 9 years ago) and `dataset` appears to work fine:

![image](https://user-images.githubusercontent.com/1702638/52725160-8d1cd900-2fa8-11e9-8c64-6c2b8187b7e9.png)

Reply:
Also tested with iphones and works fine

Reply:
https://caniuse.com/#feat=dataset

Reply:
So works in virtually all browsers including Safari except for SVG support

Seems unnecessary to roll own implementation of dataset when it's a web standard imo

Reply:
I made some tests and it works on our supported browsers thanks @garygreen ! 👍 

Fixed here: https://github.com/twbs/bootstrap/pull/23586/commits/552f2905a577ac78b55700a95200e72005839916